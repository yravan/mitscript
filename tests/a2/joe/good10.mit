// equality


// integers
x1 = 1;
x2 = 1;
x3 = -2;

print("integer equality");
print(x1 == x2); // true
print(x1 == x3); // false

// strings
s1 = "test";
s2 = "test";
s3 = "not test";

print("string equality");
print(s1 == s2); // true
print(s1 == s3); // false

// booleans
b1 = true;
b2 = true;
b3 = false;

print("boolean equality");
print(b1 == b2); // true
print(b1 == b3); // false


// records
r1 = {a1:1; a2:2; a3:3;};
r2 = {a1:1; a2:2; a3:3;};
r3 = r1;
r4 = {a1:1; a2:2;};
r5 = {};
r6 = {};

print("\nrecord equality");
print(r1 == r2); // false
print(r1 == r3); // true
print(r1 == r4); // false
print(r1 == r5); // false
print(r5 == r5); // true
print(r5 == r6); // false


// functions

f1 = fun() {
    g1 = fun() {};
    return g1;
};

f2 = fun() {
    g2 = fun() {};
    return g2;
};

f3a = fun() {
    global x;
    if (true) {
        return false;
    }
};
f3b = fun() {
    global x;
    if (true) {
        return false;
    }
};

f4a = fun() {
    global x;
    while(false) {
        print("false");
    }
    if (false) {
        return false;
    } else {
        print("dummy");
    }
    x = {a: 1; b: 2;};
};
f4b = fun() {
    global x;
    while(false) {
        print("false");
    }
    if (false) {
        return false;
    } else {
        print("dummy");
    }
    x = {a: 1; b: 2;};
};

a = f1;
b = a;

c = a();
d = f2();

print("\nFunction equality:");
print(a == a); // true
print(a == b); // true
print(a == c); // false
print(c == d); // false
print(f3a == f3b); // true
print(f4a == f4b); // true
