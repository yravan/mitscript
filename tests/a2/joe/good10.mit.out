integer equality
true
false
string equality
true
false
boolean equality
true
false

record equality
false
true
false
false
true
false

Function equality:
true
true
false
false
true
true
